hw/xbox/mcpx_apu.c

```diff
@@ -1468,6 +1502,22 @@ static void process_voice(MCPXAPUState *d,
         }
     }
 #endif
+
+    /*
+     * FIXME: For now, simply mix samples into a single bin for basic playback
+     */
+
+    // Find the bin with maximal volume and just use that to scale volume
+    int mix_vol = 0xfff;
+    for (int j = 0; j < 8; j++) {
+        mix_vol = MIN(mix_vol,vol[j]);
+    }
```

Comment:
This makes no sense. Why is this being done?

Reply:
Done here because I'm mixing everything into one bin

Reply:
Did you try my mixer (disabled code right above = `// Mix samples into voice bins `) and:

- accumulating the results of the 32 mixbins (at 1/32 of the volume; to get a hacky downmix - will contain multi-pass garbage though)
- or: assume a ITU downmix of the respective mixbins (basically what default GP / EP would do if I remember correctly)

For reference (ITU downmix approach):

[This code should be doing readback of the data that GP sends to EP (which is post-processed VP mixbins)](https://github.com/JayFoxRox/xqemu/blob/c14829135269952d81729d8d30d54b18d9f3c87c/hw/xbox/mcpx_apu.c#L882-L889).
I don't fully remember, but [I believe this code was also based on RE and does what EP does](https://github.com/JayFoxRox/xqemu/blob/c14829135269952d81729d8d30d54b18d9f3c87c/hw/xbox/mcpx_apu.c#L912-L933).

This kind of emulation basically skips GP effects processing and EP AC3 encoding, but still mimics the 5.1 to stereo conversion and VP mixbin layout suggested by DirectSound.